### ISSUE TITLE
Problems with new beacon functions

### ISSUE THREAD

**Issue by ndward** [ev3dev/ev3dev-lang-python#520]
ev3dev version: 4.14.61-ev3dev-2.2.2-ev3
ev3dev-lang-python version:
ii python3-ev3dev 1.2.0 all Python language bindings for ev3dev
ii python3-ev3dev2 2.0.0~beta1 all Python language bindings for ev3dev

My tests suggest that all the following functions fail when the physical distance between the IR sensor and the IR beacon is less than about 35cm: **distance()**, **heading()**, **heading_and_distance()**

I also believe the **beacon()** function also does not work as it should.

I held the beacon (channel 1) about 2m from the sensor which corresponds to about 100 as a **distance()** value then gradually moved the beacon towards the fixed sensor. As you can see below, the **distance()** reading decreased to about 21 (at a physical distance of about 35cm) and then when I moved the beacon even closer to the sensor the readings went very wrong, 100 for every reading. Moving the beacon back beyond 35cm did NOT cause the readings to become correct (they went first to very low values) until the beacon was about 1.6m from the sensor, giving distance values of around 80. Here are readings from a typical trial (I've added comments)

100 100 100 100 96 81 78 75 73 74 
69 68 62 55 51 48 46 46 47 48 
48 47 44 38 41 42 41 40 39 39 
39 37 33 31 31 28 26 26 26 26 
26 26 26 26 26 26 26 25 24 24 
24 24 24 23 25 23 22 21 21 21 
21 21 21 21 21 100 100 100 100 100   **# This is where the readings go wrong**
100 100 100 100 100 100 100 100 100 5 
100 100 100 100 100 100 100 100 5 100 
100 100 100 100 100 100 100 100 100 100 
100 14 100 100 100 100 100 100 0 0 **# Moving the beacon beyond 35cm does not fix the problem...**
1 2 2 3 3 4 4 4 4 5 
5 5 6 5 5 6 5 5 5 5 
5 6 6 6 7 7 8 9 8 9 
9 10 10 11 10 11 10 10 10 10 
10 10 11 10 10 10 10 11 100 86  **# ... until the beacon is nearly 2m from the sensor**
77 78 78 80 79 78 78 79 77 78 
81 80 78 80 78 79 75 75 77 69

After bringing the beacon too close (closer than 35cm) and causing the readings to go wrong it was also possible to get the readings to resume correct values by moving the beacon beyond 35cm and then turning the beacon off and on.

Obviously it is not acceptable for the sensor readings to go wrong in this way when the beacon gets closer than 35cm since this is likely to happen very often in various situations.

The problem was the same for the distance values returned by the **heading_and_distance()** function.

As for **heading()** (or the heading component of **heading_and_distance()**) the readings were correct when the beacon was at least 45cm from the sensor, with the heading values being negative when the beacon was to the left of the centreline as viewed from the sensor, and positive to the right. When the beacon was closer than about 45cm from the sensor, the readings went very wrong, with almost always values of -25, 0 or +25 being returned, and with the incorrect sign.

The **beacon()** function correctly toggles from False to True when the beacon button is pressed, but then True is always returned even after the beacon button is released.

I used this script to test the beacon functions **distance()**, **heading()** and **heading_and_distance()**, changing the name of the function as necessary:
```
#!/usr/bin/env python3
from ev3dev2.sensor.lego import InfraredSensor
from sys import stderr
from time import sleep

ir = InfraredSensor()

while True:
    for i in range(10):
        # Print to VS Code output panel
        print(ir.distance(), end=' ', file=stderr)
        sleep(0.5)
    print('', file=stderr) # start new line in VS Code
```

I ran this EV3-G program and could see that the EV3-G readings were apparently the same as those given by the Python **distance()** function for physical distances beyond 35cm but the EV3-G readings continued to have sensible values all the way down to zero distance, whereas the python **distance()** function fails for physical distances less than 35cm.
![beacon_distance](https://user-images.githubusercontent.com/9800891/45577810-3f2b6a00-b87e-11e8-8d6d-c97a2cbd8506.PNG)

**Comment by WasabiFan** [comment]
> The beacon() function correctly toggles from False to True when the beacon button is pressed, but then True is always returned even after the beacon button is released.

https://github.com/ev3dev/ev3dev-lang-python/issues/498

---

I am pretty sure this will all be fixed by the changes from #498.

**Comment by WasabiFan** [comment]
Fix released in `2.0.0~beta2`. To upgrade:

```
sudo apt-get update
sudo apt-get install --only-upgrade python3-ev3dev2
```
